### PR TITLE
Speed up SimSwap's `reverse2original`

### LINKED ISSUE #3: Speed up SimSwap's `reverse2original`
The bottleneck for speed isn't the SimSwap model, but the image paste back algorithm. We should speed that up to allow higher FPS: [reverse2original.py](https://github.com/sensity-ai/dot/blob/main/dot/simswap/util/reverse2original.py)

See also:
https://github.com/neuralchen/SimSwap/issues/137#issuecomment-957225825

### PR DESCRIPTION
**Pull request by AjinkyaIndulkar**
## Description

This PR replaces use of `cv2.warpAffine` with `kornia.geometry.transform.warp_affine` to run tranformations on CUDA if `use_gpu=True`

Closes #3.

#### Added:

- ➕ added `kornia` dependency

#### Updated:

- ⚡ updated `dot.simswap.util.reverse2original.reverse2wholeimage` function to use `kornia.geometry.transform.warp_affine` for affine transformation on CUDA device
- Optimize the use of devices and clean numpy and torch conversions

### CODE REVIEW

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:155]: you may be able to refactor these conditional branches by this https://pytorch.org/docs/stable/generated/torch.Tensor.to.html#torch-tensor-to

**Comment by giorgiop** [comment]
Can we do a speed comparison for before vs after?

**Comment by AjinkyaIndulkar** [comment]
> Can we do a speed comparison for before vs after?

I won't be able to do this due to hardware constraints. I was hoping @Ghassen-Chaabouni could run the speed tests.

File: dot/simswap/util/reverse2original.py
Comment (AjinkyaIndulkar) [reply dot/simswap/util/reverse2original.py:155]: that's correct. I would actually define a `device` variable which is set to `"cuda"` or `"cpu"` based on `use_gpu`.

I've added a TODO for this. For now, I tried to maintain the code style of this function.

**Review by ghassen1302 — CHANGES_REQUESTED** [review]
Please make the necessary changes to fix this error. This happened when I run `simswap` on my camera and on images.
```
reverse2original.py", line 158, in reverse2wholeimage
    torch.tensor(target_image_parsing).cuda(),
ValueError: At least one stride in the given numpy array is negative, and tensors with negative strides are not currently supported. (You can probably work around this by making a copy of your array  with array.copy().)
```

**Comment by AjinkyaIndulkar** [comment]
@ghassen1302 please remember to update `setup.cfg` as well.

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:168]: Sure

Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Done

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:90]: remove?

Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Done

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:110]: typo in the comment: **invert** the Affine transformation matrix

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:112]: can these 3 operations be compressed? is the conversions to numpy necessary?

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:118]: These operation implement the explicit inversion of the matrix cell by cell, and they are done in numpy.
I think it's worth to try to move the to pytorch and call `inv` (https://pytorch.org/docs/stable/generated/torch.linalg.inv.html#torch.linalg.inv)

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:71]: is the conversation to numpy necessary?

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:129]: can you add a comment explaining why we do this below?
```
if tgt_mask.sum() >= 5000:
```

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:122]: Fixed

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Fixed

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Fixed

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:141]: This was implemented by the original authors. They didn't add a comment explaining it. I think it means if the mask is big enough.

**Comment by AjinkyaIndulkar** [comment]
@ghassen1302 please remember to update the CHANGELOG as well.

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:140]: Comment added https://github.com/sensity-ai/dot/blob/feature/3-speedup-simswap/dot/simswap/util/reverse2original.py#L150

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Done https://github.com/sensity-ai/dot/blob/feature/3-speedup-simswap/dot/simswap/util/reverse2original.py#L136

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:12]: this? https://pytorch.org/docs/stable/generated/torch.isin.html#torch-isin

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:117]: can you make the `device` as a variable to be passed in the constructor instead?

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:132]: this operation does not depend on any data and must be moved outside of the loop

Comment (giorgiop) [reply dot/simswap/util/reverse2original.py:132]: same for any other in the loop, I think there is more

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:189]: move out of the loop

Comment (giorgiop) [reply dot/simswap/util/reverse2original.py:189]: same as any constant tensor in the `else`

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:198]: do you need the explicit conversion to `float` ?

**Review by giorgiop — CHANGES_REQUESTED** [review]

File: dot/simswap/util/reverse2original.py
Comment (AjinkyaIndulkar) [reply dot/simswap/util/reverse2original.py:12]: @Ghassen-Chaabouni what pytorch version are you using?

Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:12]: `1.9.0+cu111` the one that we use for dot

Comment (AjinkyaIndulkar) [reply dot/simswap/util/reverse2original.py:12]: i see. best to upgrade the dependency then.

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Done https://github.com/sensity-ai/dot/pull/20/commits/40051c2389b20d24592099882ebe80fa808a7428

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Done https://github.com/sensity-ai/dot/pull/20/commits/40051c2389b20d24592099882ebe80fa808a7428

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:132]: I think I removed them all from the loop https://github.com/sensity-ai/dot/pull/20/commits/40051c2389b20d24592099882ebe80fa808a7428

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:117]: I think using `device` like this is the best way. 
Do you think using the `device` variable in the `forward` function is better?

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:132]: this can still be moved out of the loop?

File: dot/simswap/util/reverse2original.py
Comment (giorgiop) [inline dot/simswap/util/reverse2original.py:134]: same with this?

**Comment by giorgiop** [comment]
@vassilispapadop can you test on the script if the quality is still OK?
@AjinkyaIndulkar can you test via webcam?

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:132]: No `mat_rev` needs to be calculated in the loop. The loop corresponds to the number of faces in the frame

Comment (giorgiop) [reply dot/simswap/util/reverse2original.py:132]: Isn't this a constant?

Comment (giorgiop) [reply dot/simswap/util/reverse2original.py:132]: I understand that the next instruction must be in the loop, but not this one

Comment (giorgiop) [reply dot/simswap/util/reverse2original.py:132]: the only dependency on `mat` is in the second instruction, right?
```
        mat_rev = torch.ones([3, 3]).to(device)
        mat_rev[0:2, :] = torch.tensor(mat).to(device)
        mat_rev[2, :] = torch.tensor([0.0, 0.0, 1.0]).to(device)
```

Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py:132]: I can rename the variable in `mat_rev = torch.linalg.inv(mat_rev)` to move the line out of the loop

Comment (giorgiop) [reply dot/simswap/util/reverse2original.py:132]: oh I missed that we are using the same bar for the inverse later. Yes, we can rename and avoid those two ops

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Done https://github.com/sensity-ai/dot/pull/20/commits/490de07df5c0a6f6cdf9f0e74183024084d319c5

File: dot/simswap/util/reverse2original.py
Comment (ghassen1302) [reply dot/simswap/util/reverse2original.py]: Done https://github.com/sensity-ai/dot/pull/20/commits/490de07df5c0a6f6cdf9f0e74183024084d319c5

**Comment by giorgiop** [comment]
Measured on a NVIDIA GeForce RTX 2070
```
main
Simswap: FPS 8.6
SimswapHQ: FPS 7.5  

this branch
Simswap: FPS 13.8 -> 1.6x speed up 
SimswapHQ: FPS 11.6  -> 1.5x speed up
```

I didn't notice any effect on qualtiy

**Comment by vassilispapadop** [comment]
@giorgiop i also didn't notice any difference with main branch. I do see some black lines on the left and top of the swapped image(randomly chosen) but i guess this is not related to speed improvements.

Screen Shot 2022-07-13 at 10 40 05 AM

**Comment by giorgiop** [comment]
OK so that is an issue already present in `main`. We can merge this PR if @Ghassen-Chaabouni is cool with it

**Comment by ghassen1302** [comment]
> OK so that is an issue already present in `main`. We can merge this PR if @Ghassen-Chaabouni is cool with it

Let me first update the CHANGELOG before merging

**Comment by vassilispapadop** [comment]
side note: i did video swap (24 seconds)
main branch ~2 minutes
this branch 57 seconds!!

**Review by ghassen1302 — APPROVED** [review]

**Pull request merged**
